### PR TITLE
#2976. Add grammar, semantics and type inference test for constants

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

**Comment by sgrekhov** [comment]
Specification was changed. converted to a draft

**Comment by sgrekhov** [comment]
Updated. Please review.

File: LanguageFeatures/Static-access-shorthand/type_inference_A02_t03.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+//               ^
+// [analyzer] unspecified
+// [cfe] unspecified
+
+  final v2 = const .new(2);
+//                 ^
+// [analyzer] unspecified
+// [cfe] unspecified
+
+  const v3 = const .id(3);
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/type_inference_A02_t03.dart:31]: So this is the case where there is no type annotation, and we have other cases for type variable types and intersection types, but I couldn't spot any tests about some of the structural types: Function types and record types. Do we have other tests covering those?

**Review by eernstg — COMMENTED** [review]
Looks good, couple of comments.

**Comment by eernstg** [comment]
Oh, I just found a couple of specification documents where 'promoted type variable' is used (seemingly) to mean 'intersection type'. No need to change that terminology, then.

File: LanguageFeatures/Static-access-shorthand/type_inference_A01_t01.dart [hunk 2]
```diff
@@ -56,6 +64,9 @@ main() {
   C<int>? c4 = .id<String>();
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Static-access-shorthand/type_inference_A01_t01.dart:64]: Thank you for noticing! Fixed.

File: LanguageFeatures/Static-access-shorthand/type_inference_A01_t05.dart [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  }
+}
+
+class C<T> {
+  T t;
+  C(this.t);
+
+  void test() {
+    if (T is int) {
+      T answer = .parse("42");
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Static-access-shorthand/type_inference_A01_t05.dart]: Thank you! Updated (including the comments).

File: LanguageFeatures/Static-access-shorthand/type_inference_A02_t03.dart (hunk 1 above)
Comment (sgrekhov) [reply LanguageFeatures/Static-access-shorthand/type_inference_A02_t03.dart:31]: Yes but not in this PR. I'll add them in the upcoming PR

**Review by sgrekhov — COMMENTED** [review]
Thank you. Updated. PTAL

File: LanguageFeatures/Static-access-shorthand/grammar_A01_t06.dart [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+/// @assertion We introduce grammar productions of the form:
+/// ```
+/// <postfixExpression> ::= ...                      -- all current productions
+///   | <staticMemberShorthand>
+///
+/// <constantPattern> ::=  ...             -- all current productions
+///   | <staticMemberShorthand>
+///
+/// <staticMemberShorthand> ::= <staticMemberShorthandHead> <selector*>
```
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Static-access-shorthand/grammar_A01_t06.dart:13]: (PR already out for that, thanks!)

File: LanguageFeatures/Static-access-shorthand/type_inference_A01_t01.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Static-access-shorthand/type_inference_A01_t01.dart:64]: Thanks!

File: LanguageFeatures/Static-access-shorthand/type_inference_A02_t03.dart (hunk 1 above)
Comment (eernstg) [reply LanguageFeatures/Static-access-shorthand/type_inference_A02_t03.dart:31]: OK!

**Review by eernstg — APPROVED** [review]
LGTM

**Pull request merged**